src/slack_model.jl

```diff
@@ -37,77 +37,111 @@ mutable struct SlackModel <: AbstractNLPModel
   model :: AbstractNLPModel
 end
 
+"""Like `SlackModel`, this model converts inequalities into equalities and bounds.
+"""
+mutable struct SlackNLSModel <: AbstractNLSModel
+  meta :: NLPModelMeta
+  nls_meta :: NLSMeta
+  model :: AbstractNLPModel
+end
 
-"Construct a `SlackModel` from another type of model."
-function SlackModel(model :: AbstractNLPModel)
-
-  ns = model.meta.ncon - length(model.meta.jfix)
-  ns == 0 && (return model)
-
-  jlow = model.meta.jlow
-  jupp = model.meta.jupp
-  jrng = model.meta.jrng
+function slack_meta(meta :: NLPModelMeta)
+  ns = meta.ncon - length(meta.jfix)
+  jlow = meta.jlow
+  jupp = meta.jupp
+  jrng = meta.jrng
 
   # Don't introduce slacks for equality constraints!
-  lvar = [model.meta.lvar ; model.meta.lcon[[jlow ; jupp ; jrng]]]  # l ≤ x  and  cₗ ≤ s
-  uvar = [model.meta.uvar ; model.meta.ucon[[jlow ; jupp ; jrng]]]  # x ≤ u  and  s ≤ cᵤ
-  lcon = zeros(model.meta.ncon)
-  lcon[model.meta.jfix] = model.meta.lcon[model.meta.jfix]
-  ucon = zeros(model.meta.ncon)
-  ucon[model.meta.jfix] = model.meta.ucon[model.meta.jfix]
-
-  meta = NLPModelMeta(
-    model.meta.nvar + ns,
-    x0=[model.meta.x0 ; zeros(ns)],
+  lvar = [meta.lvar ; meta.lcon[[jlow ; jupp ; jrng]]]  # l ≤ x  and  cₗ ≤ s
+  uvar = [meta.uvar ; meta.ucon[[jlow ; jupp ; jrng]]]  # x ≤ u  and  s ≤ cᵤ
+  lcon = zeros(meta.ncon)
+  lcon[meta.jfix] = meta.lcon[meta.jfix]
+  ucon = zeros(meta.ncon)
+  ucon[meta.jfix] = meta.ucon[meta.jfix]
+
+  return NLPModelMeta(
+    meta.nvar + ns,
+    x0=[meta.x0 ; zeros(ns)],
     lvar=lvar,
     uvar=uvar,
-    ncon=model.meta.ncon,
+    ncon=meta.ncon,
     lcon=lcon,
     ucon=ucon,
-    y0=model.meta.y0,
-    nnzj=model.meta.nnzj + ns,
-    nnzh=model.meta.nnzh,
-    lin=model.meta.lin,
-    nln=model.meta.nln,
+    y0=meta.y0,
+    nnzj=meta.nnzj + ns,
+    nnzh=meta.nnzh,
+    lin=meta.lin,
+    nln=meta.nln,
   )
+end
+
+"Construct a `SlackModel` from another type of model."
+function SlackModel(model :: AbstractNLPModel)
+  model.meta.ncon == length(model.meta.jfix) && return model
+
+  meta = slack_meta(model.meta)
 
   return SlackModel(meta, model)
 end
 
+function SlackNLSModel(model :: AbstractNLSModel)
+  ns = model.meta.ncon - length(model.meta.jfix)
+  ns == 0 && return model
+
+  meta = slack_meta(model.meta)
+  nls_meta = NLSMeta(model.nls_meta.nequ,
+                     model.meta.nvar + ns,
+                     [model.meta.x0; zeros(ns)])
+
+  return SlackNLSModel(meta, nls_meta, model)
+end
+
+const SlackModels = Union{SlackModel,SlackNLSModel}
+
 import Base.show
 # TODO: improve this!
 # show(nlp :: SlackModel) = show(nlp.model)
 
 # retrieve counters from underlying model
 for counter in fieldnames(Counters)
   @eval begin
-    $counter(nlp :: SlackModel) = $counter(nlp.model)
+    $counter(nlp :: SlackModels) = $counter(nlp.model)
```

Comment:
Why are we saying `SlackModels` instead of `SlackModel` now?

Reply:
`SlackModels=Union{SlackModel,SlackNLSModel}`

Reply:
Sorry, I missed it.